CAMEL-14385: add a camel-cron component

nicolaferraro opened the pull request:
Fix CAMEL-14385

Refs: https://github.com/apache/camel-k-runtime/pull/223

oscerd commented on components/camel-cron/src/test/resources/log4j2.properties:
```diff
@@ -0,0 +1,33 @@
+rootLogger.level = INFO
+rootLogger.appenderRef.file.ref = file
+
+#rootLogger.appenderRef.stdout.ref = stdout
```
We can remove those I believe

nicolaferraro commented:
Thanks @oscerd, I've checked the spring-boot starter with dev profile and it seems to work. I guess it's cannot be currently merged, right?

oscerd commented:
> Thanks @oscerd, I've checked the spring-boot starter with dev profile and it seems to work. I guess it cannot be currently merged, right?

Yeah, we'll need to create the starters once we release 3.1.0, but it's already a good start :-)

davsclaus commented on components/camel-cron/pom.xml:
```diff
@@ -0,0 +1,84 @@
+
+        <!-- requires camel-core -->
+        <dependency>
+            <groupId>org.apache.camel</groupId>
```
What makes it require camel-core, we should ideally only depend on camel-support

davsclaus commented on components/camel-cron/src/main/java/org/apache/camel/component/cron/CronEndpoint.java:
```diff
@@ -0,0 +1,105 @@
+
+    @Override
+    public Consumer createConsumer(Processor processor) throws Exception {
+        return delegate.createConsumer(processor);
```
You need also to do

configureConsumer(consumer)

davsclaus commented on components/camel-cron/src/main/java/org/apache/camel/component/cron/CronEndpoint.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    @Override
+    public void setSynchronous(boolean synchronous) {
```
Ah yeah we may need to do this on other delegate endpoints (create a jira ticket)

davsclaus commented on components/camel-cron/src/main/java/org/apache/camel/component/cron/CronHelper.java:
```diff
@@ -0,0 +1,65 @@
+            return service;
+        }
+
+        // Fallback to service loader
```
In Camel we have FactoryFinder as our way of loading these kind of stuff. It made it work in OSGi too. However maybe service loader is okay in these modern cloud days.

lburgazzoli commented on components/camel-quartz/src/main/java/org/apache/camel/component/quartz/cron/CamelQuartzCronService.java:
```diff
@@ -0,0 +1,72 @@
+        String uri = uriPath + "?" + query;
+
+        QuartzComponent quartz = context.getComponent("quartz", QuartzComponent.class);
+        return quartz.createEndpoint(uri);
```
Wonder if after we get https://issues.apache.org/jira/browse/CAMEL-14389 done we should start using the type-safe dsl for scenario like this one as it would ease refactoring/deprecations

lburgazzoli commented on components/camel-cron/src/main/java/org/apache/camel/component/cron/CronComponent.java:
```diff
@@ -0,0 +1,113 @@
+ * The Camel cron component.
+ */
+@Component("cron")
+public class CronComponent extends DefaultComponent {
```
Maybe we should implement `DelegateEndpoint`

nicolaferraro replied:
You mean.. providing kinda DelegateEndpointSupport?

nicolaferraro commented:
Thanks for the hints. I shoud have fixed the code.

I've added another implementation in camel-spring, so that it can be used in Spring or Spring-boot oob without pulling quartz.

I've switched to FactoryFinder as well.
